src/psyclone/psyir/transformations/arrayrange2loop_trans.py

```diff
@@ -214,6 +227,40 @@ def validate(self, node, options=None):
                             f"assignment '{node}'.")
                     break
 
+        if not options:
+            options = {}
+        allow_string_array = options.get("allow_string", False)
+        # If we allow string arrays then we can skip the check.
+        if allow_string_array:
+            return
+        lhs = node.lhs
+        # ArrayMixin datatype lookup can fail if the indices contain a
+        # Call or Intrinsic Call. We catch this exception and continue
+        # for now - TODO #1799
+        try:
+            if lhs.datatype.intrinsic == ScalarType.Intrinsic.CHARACTER:
+                raise TransformationError(
+                    "The ArrayRange2LoopTrans transformation doesn't allow "
+                    "character arrays by default. This can be enabled by "
+                    "passing the allow_string option to the transformation."
+                )
+        except NotImplementedError:
+            pass
+        for child in node.rhs.walk((Literal, Reference)):
+            # Skip over indices
+            if child.ancestor(ArrayReference) is not None:
+                continue
+            try:
+                if child.datatype.intrinsic == ScalarType.Intrinsic.CHARACTER:
```

Comment:
Some considerations (skip Unresolved, Unsupported and drop previous continue condition) as the NEMO version of the loop also apply here.

Reply:
Changed